### ISSUE TITLE
Observable.tailRecM leaks memory when inner loop attaches callbacks to `onNext`

### ISSUE THREAD

**Issue by sirmax** [monix/monix#791]
This problem was initially mentioned in gitter – https://gitter.im/monix/monix?at=5c0a30de5e409525032ac596

Consider this simple infinite sequence
```
def loop = Observable.tailRecM(0) { i =>
  val out = Observable(Left(i + 1)).delayExecution(1.second)
  out.dump(s"tailRecM($i)")
}
```

While the loop itself can be considered tail-recursive – it emits only one `Left(…)` as the last element – running it will eventually end up with OOME.

The leak in this example comes from `out.dump(…)` . `DumpObservable` attaches a callback to `onNext`
```
subscriber.onNext(elem).syncOnStopOrFailure(_ => downstreamActive.cancel())
```
and this callback leaves hanging forever despite the fact that we're probably not interested in it after emitting `onComplete`.

Important: `DumpObservable` is mentioned here only as an example, in fact any other observable that attaches callbacks to `onNext` (e.g. `ConcatMapObservable`) would produce the similar leak.

IMO, if the fix is possible at all, it should be done in `TailRecMObservable`.  Perhaps it's possible to return a `CancelableFuture` in response to `onNext(Left(…))` and then cancel it if upstream calls `onComplete`.

**Comment by sirmax** [comment]
BTW, maybe it's OK to keep `tailRecM` unchanged and provide a stricter alternative that allows at most one `onNext(Left(…))` and simply responds with `Stop` to it.

**Comment by alexandru** [comment]
We'll have to investigate.

```
subscriber.onNext(elem).syncOnStopOrFailure(_ => downstreamActive.cancel())
```

We register a listener indeed, however the question is why do we keep a reference to that `Future[Ack]` around.

**Comment by sirmax** [comment]
> why do we keep a reference to that `Future[Ack]` around

That's because `subscriber.onNext(elem): Future[Ack]` in case of infinite recursion is, basically, the end of the whole steam which never completes.

**Comment by sirmax** [comment]
In other words, `tailRecM` does not impose any restrictions upon the user's code, letting it emit another `Either[A, B]` after `Left(a)`. To do so `onNext(Left(…))` returns a `Future[Ack]` that completes when the whole nested sub-stream completes, no matter how deep it is. In case of recursive loop it's infinitely deep.

**Comment by alexandru** [comment]
Indeed, however via the `Observable` protocol it is legal to do this:

```scala
out.onNext(Left(...))
// No need to wait on the acknowledgement of the last onNext
out.onComplete()
```

So the final `onComplete` doesn't necessarily have to wait on the last `onNext`, we only wait when it isn't very convenient or safe to do otherwise. When developing `tailRecM` I know I took this into account. 

1. Does the issue manifest without `dump`? 
2. Does it manifest without `delayExecution`?

**Comment by sirmax** [comment]
> So the final `onComplete` doesn't necessarily have to wait on the last `onNext`

And it doesn't wait – `onComplete` gets called right after `onNext`. But the problem is that callbacks are already attached to the "root" `Future[Ack]` and remain there.

> Does the issue manifest without `dump`?

It manifests with any other combinator that attaches listeners. Initially I got this problem with inner `flatMap`, not `dump`.

> Does it manifest without `delayExecution`?

Will have to test it somehow without melting my CPU :)

**Comment by alexandru** [comment]
I see. It's a problem indeed.

**Comment by sirmax** [comment]
Checked without `delayExecution`. Got about 15K strong-reachable `dump` callbacks in a few seconds. So yeah, it leaks.
